doc/developer/design/20230531_system_privileges.md

```diff
@@ -0,0 +1,97 @@
+privilege via `GRANT ...` and you destroy them via `REVOKE ...`.
+
+Role attributes present some problems as they currently exist:
+
+- It can be confusing that there are two separate privilege primitives that control a role's
+  privilege, and they have different semantics.
+- Role attributes are not inherited which makes it cumbersome to grant attributes to a large
+  number of roles.
+- Users are not able to easily grant the ability to create top level objects to many current and
+  future users in a short amount of time.
+
+## Goals
+
+- Reduce confusion around privileges.
+- Allow the ability to create databases, clusters, and roles to be inherited through role
+  membership.
+- Allow users to easily grant the ability to create top level objects to many current and future
+  users in a short amount of time.
+- Maintain PostgreSQL compatibility.
+
+## Overview
+
+We will add a new type of privilege called System Privileges. They will look and act just like
+normal object privileges, but they will grant the abilities that would otherwise be granted through
+role attributes. They will exist in parallel with role attributes.
+
+NOTE: Other databases, such as CockroachDB and Snowflake, have some form of system privileges that
+control similar privileges to role attributes.
+
+## Detailed description
+
+System privileges will be granted with the following syntax:
+`GRANT <privilege-specification> ON SYSTEM TO <role-specification>`
+
+System privileges will be revoked with the following syntax:
+`REVOKE <privilege-specification> ON SYSTEM FROM <role-specification>`
+
+System privileges and role attributes will interact in the following ways:
+
+- Granting a system privilege will grant the privilege AND set the attribute to true.
+- Revoking a system privilege will revoke the privilege AND set the attribute to false.
+- Altering a role attribute will affect the role attribute but will not change the system privilege.
+- When checking to see if a role is allowed to execute a command we will check if they have the
+  system privilege or the role attribute.
+
+This approach has the following nice properties:
+
+- The ability to use system privileges without thinking about role attributes.
+- The ability to use role attributes without thinking about system privileges.
+- System privileges always take priority over role attributes when using both.
+
+We will support the following system privileges:
+
+- `CREATEDB`: Allows users to create databases.
+- `CREATECLUSTER`: Allows users to create clusters.
+- `CREATEROLE`: Allows users to create, modify, and delete roles.
+
+System privileges are inherited through role membership.
+
+Whenever someone alters a role attribute we will issue a warning indicating that system privileges
+take precedence and are preferred to role attributes. Additionally, the documentation will focus on
+system privileges and indicate that role attributes are there for PostgreSQL compatibility and not
+the preferred approach.
```

Comment:
Working towards the goal of reducing confusion, is it possible to add an explain statement or a function that covers role attributes and system privileges? e.g. something like
```
EXPLAIN PRIVILEGES joe;

      create_db       |      create_cluster      |      create_role     
--------------------------------------------------------------------------
 role=no,sys_priv=yes |   role=no,sys_priv=no    |  role=yes,sys_priv=yes
```

Reply:
I think the following function would fit in nicely with our existing RBAC functions (https://materialize.com/docs/sql/functions/#access-privilege-inquiry-func):
`has_system_privilege([role: text or oid, ]privileges: text) -> bool`

I also think at a future point we should add a `SHOW` commands for all privileges of the following format:
`SHOW PRIVILEGES [ON <type> <name>]`
for example:
- `SHOW PRIVILEGES` would show all privileges for the current user.
- `SHOW PRIVILEGES ON SYSTEM` would show all system privileges for the current user.
- `SHOW PRIVILEGES ON TABLE t` would show all table privileges for the current user on `t`. 
- etc.

I'll add these to the doc.

Reply:
Great idea! I think `SHOW PRIVILEGES [ON <type> <name>]` is the right syntax here. We can file a ticket for this as a follow up task after launch.